Add String.ToLower() and String.ToUpper()

caesay opened the pull request:
I think there might be something wonky with the C implementation, so you might want to check that before merging. I can't get the tests to run properly on my machine as of yet.

pfusik commented on GenC.fu:
```diff
@@ -2057,6 +2059,18 @@ public class GenC : GenCCpp
+			Include("string.h");
+			Include("ctype.h");
+			this.StringToLower = true;
+			WriteCall("FuString_ToLower", obj, args[0], args[1]);
```
WriteGlib("utf8_strdown(");
    WriteTemporaryOrExpr(obj, FuPriority.Argument);
    Write(", -1)");
    break;

pfusik commented on GenCpp.fu:
```diff
@@ -808,6 +808,26 @@ public class GenCpp : GenCCpp
+			obj.Accept(this, FuPriority.Argument);
+			Write("; ");
+			Write("std::transform(data.begin(), data.end(), data.begin(), ");
+			Write("[](unsigned char c) { return std::tolower(c); }); ");
```
This is not Unicode-aware.

pfusik commented on test/StringLower.fu:
```diff
@@ -0,0 +1,9 @@
+	{
+		string() s = "MiXeD cAsE";
+        string() n = s.ToLower();
+        return n == "mixed case" && s == "MiXeD cAsE";
```
Please also test on `string` reference.

pfusik commented on test/StringLower.fu:
```diff
@@ -0,0 +1,9 @@
+{
+	public static bool Run()
+	{
+		string() s = "MiXeD cAsE";
```
string() s = "MiXeD cAsE"; //FAIL: cl

Because OpenCL doesn't support heap allocations, so no string storage.

pfusik commented on test/StringLower.fu:
Rename to `StringToLower.fu`.

pfusik reviewed (changes requested):
Thank you for your contribution!
My main concern is that the C and C++ implementations are not Unicode-aware. For C, there are GLib functions. C++ probably needs some external dependency.
Also, please run `make` before committing (`libfut.js` is missing) and document in `doc/reference.md`.

pfusik commented on test/StringUpper.fu:
```diff
@@ -0,0 +1,9 @@
+	public static bool Run()
+	{
+		string() s = "MiXeD cAsE";
+        string() n = s.ToUpper();
```
Please indent with tabs.

caesay commented:
I have incorporated your suggestions, and rebased to 8edec9b3596e1c7cb5db551b6da0e3ba1c90bb5e. The only thing I couldn't do is fix the unicode C++ support. I would suggest that we include https://github.com/sheredom/utf8.h in the hpp output (similar to how I suggested outputting subprocess.h and json.h) and use  utf8lwr etc, but there's no precedent for how to do this in fut at the moment. I guess we need to store it as a string somewhere and include it if necessary.

pfusik reviewed (approved):
LGTM, modulo C++ which I'll handle myself. Thanks!

caesay commented:
I tried utf8.h yesterday and agree, ran into some issues with it. ICU looks like a very big / complex dependency to try and include for a library I feel. You have to link in it's static libs as well as bring in it's header files. A lot of the appeal of fusion for me comes from the fact that the output is very standalone and can just be dropped into another project as a source file. Also at the moment afaik, vckpg only supports source-only libraries, so if making a library using fusion and distributing it via vcpkg will only be possible without ICU. 

Headers like `json.h` or `subprocess.h` are great because they can be prepended to the header that is produced by fusion, I saw on a StackOverflow post that boost's unicode string library is header only, but it is just an abstraction, and relies on a configurable backend.

If we included boost.locale header, by default we could use the `winapi` backend on windows, and `posix` on windows. Since boost also supports ICU, we could add support for the consumer of the library to _optionally_ rely on ICU instead. This keeps our fusion output fairly portable but also allows for "proper" unicode support at the consumers discretion.

caesay commented:
Also see https://www.boost.org/doc/libs/1_54_0/libs/locale/doc/html/using_localization_backends.html

pfusik commented:
I have no experience with Boost, ICU or vcpkg. I made a decision based on https://stackoverflow.com/a/24063783/2032514
Unicode case manipulation is no trivial topic and I don't want to reinvent the wheel by maintaining my own implementation.

I got ICU on Windows with:

    pacman -S mingw-w64-x86_64-icu

I suppose it's not harder on Linux or macOS. Doesn't `vcpkg` provide the ICU package?

caesay commented:
Speaking of proper Unicode support - I looked at that issue you linked, which also mentioned that std::string->substr is unaware of utf-8 multi-byte sequences, so could end up cutting a character in half. This is not consistent with the C# behavior, which uses UTF-16 for example, so Substring always acts on characters, not bytes.

pfusik commented:
In Fusion it's not "characters", but _code units_. Just as in UTF-8 one _code point_ can be stored in several bytes, in UTF-16 one code point can be stored as two values (_surrogate pair_).
